### DISCUSSION TITLE
detection/tools/smoother/

### DISCUSSION THREAD

**Discussion by giscus** [roboflow/supervision#937] [Docs]
# detection/tools/smoother/

A set of easy-to-use utilities that will come in handy in any computer vision project.

https://supervision.roboflow.com/detection/tools/smoother/

**Comment by tobieabel** [comment]
Really pleased the tracking functionality is getting additional support, its almost as important as model accuracy in real world applications.
Just tried out the Smoother and I noticed that when no detections were found it replaced the detections tracker ID empty list with a "None" object.  Not sure if thats expected behaviour, but it caused errors when trying to loop though the tracker ID array.

**Comment by SudhanshuBlaze** [comment]
This is my code using Detection Smoother for classwise traffic counting using “line zone” , the confusing thing is the supervision is able to detect the cars, persons and bikes, but when they cross the line zone, the count is not always updated. For example, if 5 cars cross the line zone then count only increases by 2. I also want the count to update vehicle class wise.

```
import supervision as sv
from ultralytics import YOLO
import cv2
import numpy as np

# Load YOLO model
model = YOLO("yolov9e.pt")
tracker = sv.ByteTrack()
frames_generator = sv.get_video_frames_generator("resources/741755_Kazakhstan Traffic Cars Road_By_Danil_Nevsky_Artlist_HD.mp4")
video_info = sv.VideoInfo.from_video_path("resources/741755_Kazakhstan Traffic Cars Road_By_Danil_Nevsky_Artlist_HD.mp4")

# Define the line for counting
start, end = sv.Point(x=50, y=300), sv.Point(x=1000, y=250)
line_zone = sv.LineZone(start=start, end=end)

# Initialize counters
crossed_count_in = 0
crossed_count_out = 0

smoother = sv.DetectionsSmoother()

bounding_box_annotator = sv.BoundingBoxAnnotator()

with sv.VideoSink("supervision_output.mp4", video_info=video_info) as sink:
    for frame in frames_generator:
        result = model(frame)[0]
        detections = sv.Detections.from_ultralytics(result)
        detections = tracker.update_with_detections(detections)
        detections = smoother.update_with_detections(detections)

        crossed_in, crossed_out = line_zone.trigger(detections)
        if np.any(crossed_in):
            crossed_count_in += np.sum(crossed_in)
        if np.any(crossed_out):
            crossed_count_out += np.sum(crossed_out)

        annotated_frame = bounding_box_annotator.annotate(frame.copy(), detections)

        # Annotate the line on the frame
        cv2.line(annotated_frame, (start.x, start.y), (end.x, end.y), (0, 0, 255), 2)  # Red line

        # Annotate counts on the frame
        cv2.putText(annotated_frame, f"Crossed In: {crossed_count_in}", (50, 50), cv2.FONT_HERSHEY_SIMPLEX, 1, (0, 255, 0), 2)
        cv2.putText(annotated_frame, f"Crossed Out: {crossed_count_out}", (50, 100), cv2.FONT_HERSHEY_SIMPLEX, 1, (0, 255, 0), 2)

        sink.write_frame(annotated_frame)
```

**Reply by LinasKo** [reply]
Hi @SudhanshuBlaze 👋 

Lately we've been recommending folks to try `pip install supervision==0.21.0.rc5` - it fixes an issue with the tracker and there a **small** chance that it would work reliably in your case. Could you please try that?
